_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+
+One can get statuses from CodePipeline by using AWS CLI or AWS SDK which is available for many different languages. AWS [CloudWatch](https://aws.amazon.com/cloudwatch/) alarms and metrics are also accessible via CLI and SDK. 
+
+One way to extract data would be to make an AWS user to every account, generate a access key id and a secret key and poll end points directly from the radiator software. A variant of this would be to make a AWS role and configure cross account permissions to switch to that role. 
+
+Third way which I chose is to make a Lambda function which exposes statuses in one JSON data response. That way it is possible to get all necessary data from one endpoint. The endpoint should be protected with an api key which is supported quite nicely in the Amazon [API Gateway](https://aws.amazon.com/api-gateway/). The radiator should poll the endpoints and generate a view of the account status.
+
+AWS Lambda environment currently supports Node.js, Java, C# and Python languages. For an endpoint exposer I chose Python which have excellent [Boto 3](https://boto3.readthedocs.io/en/latest/) SDK for accessing AWS.
+
+First I started from what I want as a result. The result should be JSON response with different keys for alarms, pipelines and metrics. 
+
+```json
+{
+    "alarms": [
+        {
+            "AlarmName": "product-name-service-500",
+            "StateValue": "OK"
+        }
+    ],
+    "pipelines": [
+        {
+            "name": "ProjectInfraBuildPipeline",
+            "currentStatus": "Succeeded"
+        }
+    ],
+    "metrics": [
+        {
+            "name": "RDS CPU Utilization",
+            "statistics": "Average",
+            "unit": "Percent",
+            "result": 7.004
+        }
+    ]
+}
+```
+
+The main handler is therefore quite simple. There is no need to use any parameters because we return always all the pipelines and alarms.   
+
+```python
+def status(event, context):
+    result = {
+        "alarms":get_alarms(),
+        "pipelines" : get_pipelines(),
+        "metrics" : get_metrics()}
+    return {"body": json.dumps(result)}
+```
+
+Getting pipeline statuses and alarms is quite straightforward. The alarms can be fetched with one API call. Pipelines need to be first fetched and then current statuses can be checked one by one. For the radiator I also formatted the result a little bit.
+
+Metrics cannot be fetched with one call because AWS has tons of different kind metrics available. Currently I make a Python data structure which describes what kind of metric I want to expose. The next code example gets average RDS CPU utilization of last 10 minutes.
```

Comment:
"tons of different kinds of metrics"

"CPU utilization from the last 10 minutes"